### PR TITLE
Add run function to the MitoAnalysis class

### PR DESCRIPTION
**Pull request by marthacryan**
# Description
Adds a function to the `Mito Analysis` class that takes no arguments and re-runs the analysis with it's original parameter values. Throws a `NotImplementedError` for `df_name` type parameters, because the initial values won't be stored in the `ParamMetadata`. 

## Questions
* What name should the function default to? 
  * Current implementation: "automate"
* Why is the file path string given from `get_parameterizable_params` as `initial_value` an r-string? I had to manually remove that from the string to run it.
  * I'm assuming this is because it's supposed to be run as transpiled code and not as an argument to an *actual* function

# Testing
Adds a test file to `tests/streamlit` called `test_mito_analysis.py`. It tests that, when given code and param metadata, the proper output is generated. It does this by adding a column with a formula in the given code, and checking that it's in the output of the function. 

# Documentation
Should add documentation to the streamlit docs once we add custom parameters.

### CODE REVIEW

**Review by naterush — CHANGES_REQUESTED** [review]
This is freaking awesome. The ease with which you can re-run this on new datasets: so cool, and absolutely what a developer on Streamlit (me, 10x times) want. 

My only major thing I think we should change is around *args and **kwargs. In generally, I think it's totally legit to pass everything by **kwargs, but I also feel that _positional arguments that were passed to the spreadsheet__ should be passable through *args. <- to be clear, this is different than my comment below, where I say they should be required to be *args.

To make sure what I'm communicating is clear, I think the following would be great test cases for this new behavior!

## What should be allowed

### Pass initial params through args
```
analysis = spreadsheet(df)
analysis.run(df)
```

### Pass initial params through keywords
```
analysis = spreadsheet(df)
analysis.run(df1=df)
```
Or if you set df_names:
```
analysis = spreadsheet(df, df_names=['first_dataframe'])
analysis.run(first_dataframe=df)
```


## What should error

### Pass both keyword and args

```
analysis = spreadsheet(df)
analysis.run(df, df1=df)
```
Error: TypeError: run() got multiple values for argument 'df1'

### Don't pass keyword args or args
```
analysis = spreadsheet(df)
analysis.run()
```
Same error we have currently.

File: mitosheet/mitosheet/streamlit/v1/spreadsheet.py [hunk 1]
```diff
@@ -112,11 +112,38 @@ def get_function_from_code_unsafe(code: str) -> Optional[Callable]:
...
+        return self.__param_metadata
+
+    def run(self, **kwargs):
+        params = {}
+
+        # Error handling for required arguments
+        required_args = [param['name'] for param in self.__param_metadata if param['required']]
+        for required_arg in required_args:
+            if required_arg not in kwargs.keys():
+                raise TypeError(f'MitoAnalysis.run() missing required argument {required_arg}')
```
[the remark this answers is not in the document]
Comment (marthacryan) [reply mitosheet/mitosheet/streamlit/v1/spreadsheet.py]: Updated it! Lmk if you think the wording could be better though

File: mitosheet/mitosheet/api/get_parameterizable_params.py [hunk 2]
```diff
@@ -44,7 +44,7 @@ def map_tuple_to_param_data(args):
                 param = args[0]
                 param_name = args[1] 
                 return {
-                        'initial_value': param[0],
+                        'initial_value': param[0][2:-1],
```
Comment (naterush) [inline mitosheet/mitosheet/api/get_parameterizable_params.py:47]: Wanna leave a comment here?

**Review by naterush — CHANGES_REQUESTED** [review]
## Tests
Can you add a test for MitoAnalysis in the presence of:
1. custom importers
2. custom sheet functions
3. custom editors

## Bug in dynamic imports?

Code looks good, but I'm having an error in the streamlit app you pushed. I think there's a few things:

1. First, I needed to pass the df to the `.run`. Easy fix.
2. Then, get this weird importing error. Think it's with how we're detecting import lines, namely, I think need to change the line in do dynamic imports to:
```
    import_lines = [line.strip() for line in code.split("\n") if line.strip().startswith(("from ", "import "))]
```

Note the extra spaces so we don't detect lines that have a dataframe that start with the word import. 

But even then, the function doesn't run properly. Not sure what's up here, wanna dig in and figure it out? (and maybe write a good test if you catch a bug).

**Comment by naterush** [comment]
This looks good to merge. Let's merge it once we get this released fixed up. 

Feel free to branch off this for now, while we're finishing up the release.

**Pull request merged**